distsys/resources/tcpmailboxes.go

```diff
@@ -0,0 +1,394 @@
+package resources
+
+import (
+	"encoding/gob"
+	"fmt"
+	"github.com/UBC-NSS/pgo/distsys"
+	"io"
+	"net"
+	"time"
+)
+
+// Mailboxes as Archetype Resource
+// -------------------------------
+
+const (
+	tcpNetworkBegin = iota
+	tcpNetworkValue
+	tcpNetworkPreCommit
+	tcpNetworkCommit
+)
+
+type TCPMailboxKind int
+
+const (
+	TCPMailboxesLocal = iota
+	TCPMailboxesRemote
+)
+
+const (
+	tcpMailboxesReceiveChannelSize          = 100                   // TODO: this should be a configuration option
+	tcpMailboxesTCPTimeout                  = 1 * time.Second       // TODO: same as above
+	tcpMailboxesReadTimeout                 = 20 * time.Millisecond // TODO: same
+	tcpMailboxesConnectionDroppedRetryDelay = 50 * time.Millisecond // TODO: same
+)
+
+// AddressMappingFn is responsible for translating the index, as in network[index] from distsys.TLAValue to a pair of
+// TCPMailboxKind and address string, where the address string would be appropriate to pass to net.Listen("tcp", ...)
+// or net.Dial("tcp", ...). It should return TCPMailboxesLocal if this node is to be the only listener, and it should
+// return TCPMailboxesRemote if the mailbox is remote and should be dialed. This could potentially allow unusual setups
+// where a single process "owns" more than one mailbox.
+type AddressMappingFn func(distsys.TLAValue) (TCPMailboxKind, string)
+
+// TCPMailboxesArchetypeResourceMaker produces a distsys.ArchetypeResourceMaker for a collection of TCP mailboxes.
+// Each individual mailbox will match the following mapping macro, assuming exactly one process "reads" from it:
+//
+//     mapping macro TCPChannel {
```

Comment:
It's good to update the mapping macro with the one that also considers process failures.

Reply:
But this implementation doesn't actually consider them. I think that (and various other things) might be better as a future change, once the implementation catches up.

Reply:
Underlying TCP connection actually considers process failures.

Reply:
Sure, TCP does, but this protocol doesn't. It just panics under some cases, rather than attempting to re-establish a connection or something.

There's even a note below that some crash recovery cases are stubbed... nothing like that has ever been tested, and I don't expect it to work.

That's why I think that adding explicit support / implementation of your ideas is a second changeset here. It would look strange for the docs to have all this extra mechanism, but for the code to fail under those conditions.

Reply:
My point is not related to crash-recovery failure. I'm just saying that if a process is failed, you cannot send a message to it and it's good to have this explicitly in the mapping macro.

Reply:
I see... yes, that makes sense. I didn't think of it like that. I'll edit the docs (and include the expected format of `$variable`).